algorithms/src/sorting/impl/Kokkos_SortByKeyImpl.hpp

```diff
@@ -0,0 +1,322 @@
+
+#if defined(KOKKOS_ENABLE_ROCTHRUST)
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties, class... MaybeComparator>
+void sort_by_key_hipthrust(
+    const Kokkos::HIP& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  const auto policy = thrust::cuda::par.on(exec.cuda_stream());
+  auto keys_first   = ::Kokkos::Experimental::begin(keys);
+  auto keys_last    = ::Kokkos::Experimental::end(keys);
+  auto values_first = ::Kokkos::Experimental::begin(values);
+  thrust::sort_by_key(policy, keys_first, keys_last, values_first,
+                      std::forward<MaybeComparator>(maybeComparator)...);
+}
+#endif
+
+#if defined(KOKKOS_ENABLE_ONEDPL) && \
+    (ONEDPL_VERSION_MAJOR > 2022 ||  \
+     (ONEDPL_VERSION_MAJOR == 2022 && ONEDPL_VERSION_MINOR >= 2))
+template <class KeysDataType, class... KeysProperties, class ValuesDataType,
+          class... ValuesProperties, class... MaybeComparator>
+void sort_by_key_onedpl(
+    const Kokkos::SYCL& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  if (keys.stride(0) != 1 && values.stride(0) != 1) {
+    Kokkos::abort(
+        "SYCL sort_by_key only supports rank-1 Views with stride(0) = 1.");
+  }
+
+  // Can't use Experimental::begin/end here since the oneDPL then assumes that
+  // the data is on the host.
+  auto queue  = exec.sycl_queue();
+  auto policy = oneapi::dpl::execution::make_device_policy(queue);
+  const int n = keys.extent(0);
+  oneapi::dpl::sort_by_key(policy, keys.data(), keys.data() + n, values.data(),
+                           std::forward<MaybeComparator>(maybeComparator)...);
+}
+#endif
+
+template <typename ExecutionSpace, typename PermutationView, typename ViewType>
+void applyPermutation(ExecutionSpace const& space,
+                      PermutationView const& permutation,
+                      ViewType const& view) {
+  static_assert(std::is_integral<typename PermutationView::value_type>::value);
+
+  auto view_copy = Kokkos::create_mirror(space, view);
+  Kokkos::deep_copy(space, view_copy, view);
+  Kokkos::parallel_for(
+      "Kokkos::sort_by_key_via_sort::permute_" + view.label(),
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, view.extent(0)),
+      KOKKOS_LAMBDA(int i) { view(i) = view_copy(permutation(i)); });
+}
+
+template <class ExecutionSpace, class KeysDataType, class... KeysProperties,
+          class ValuesDataType, class... ValuesProperties,
+          class... MaybeComparator>
+void sort_by_key_via_sort(
+    const ExecutionSpace& exec,
+    Kokkos::View<KeysDataType, KeysProperties...>& keys,
+    Kokkos::View<ValuesDataType, ValuesProperties...>& values,
+    MaybeComparator&&... maybeComparator) {
+  auto const n = keys.size();
```

Comment:
What if `keys.size() != values.size()`? We may be able to check this for some data types at compile time (i.e. static extents). As for runtime I think we should assert if the sizes are invalid to avoid messy errors when we permute. This applies to all the different sort variants and we could check it in `sort_by_key`

Reply:
nvm was in recent changes, but you could still consider the static check though it's not absolutely necessary

Reply:
Could you please give a snippet of the check you are thinking about? 
I'm not too familiar with doing static checks. 
One could also try to do something like that for the regular `sort`.

Reply:
Something like (may be wrong I haven't written it out):

```cpp
using key_view = Kokkos::View<KeysDataType, KeysProperties...>;
using value_view = Kokkos::View<ValuesDataType, ValuesProperties...>;
static_assert(key_view::static_extent(0) == 0 || value_view::static_extent(0) == 0 || key_view::static_extent(0) == value_view::static_extent(0));
```
The 0s representing a dynamic extent